src/backends/tensorrt/plan_backend.cc

```diff
@@ -801,6 +824,56 @@ PlanBackend::Context::InitializeConfigOutputBindings(
   return Status::Success;
 }
 
+Status
+PlanBackend::Context::PeekShapeTensor(
+    const InferRequestHeader::Input& input, const Scheduler::Payload& payload,
+    std::vector<int64_t>* shape)
+{
+  // It is the caller's responsibility to only call on shape tensors,
+  // which means the datatype must be INT32.
+  int64_t element_cnt = GetElementCount(input.dims());
+  size_t content_byte_size =
+      element_cnt * GetDataTypeByteSize(DataType::TYPE_INT32);
+  const char* content;
+
+  // Get the tensor contents into contiguous CPU memory...
+  std::unique_ptr<AllocatedSystemMemory> contiguous_buffer;
+  bool cuda_copy = false;
+  RETURN_IF_ERROR(GetContiguousInputContent(
+      input.name(), TRTSERVER_MEMORY_CPU, 0 /* src_memory_type_id */, payload,
+      &content, &content_byte_size, &contiguous_buffer, &cuda_copy));
+
+  if (cuda_copy) {
+    cudaStreamSynchronize(stream_);
+  }
+
+  shape->clear();
+
+  const int32_t* dims = reinterpret_cast<const int32_t*>(content);
+  for (int64_t i = 0; i < element_cnt; ++i) {
+    shape->push_back(dims[i]);
+  }
+
+  // If have an provider override for this tensor already (because
+  // peek was already called, then mark it as not being consumed so
+  // that the next peek or usage will be able to access it. If don't
+  // already have an override, add it...
+  if (payload.request_provider_->HasInputOverride(input.name())) {
+    payload.request_provider_->SetInputOverrideConsumed(input.name(), false);
+  } else {
+    InferRequestProvider::InputOverride override;
+    override.content_.assign(dims, dims + element_cnt);
```

Comment:
So basically, we are using InputOverride() like this to reset the data in the buffer?

Reply:
InputOverride is the way you tell the provider to use the "override" data instead of using the data provided by the inference request.  We don't want to have to move the data from the GPU everytime it is peeked so we use the override to cache it.